File: ocaml/bf.ml
open Str;;

let out_start = "extern int putchar(int);\nextern char getchar();\n\nchar array[30000] = {0}; char *ptr = array;\nint main (int argc, char *argv[]) {"

let (mod) x y = ((x mod y) + y) mod y;;

let read_file filename =
        let ch = open_in filename in
        let s  = really_input_string ch (in_channel_length ch) in
        close_in ch;
        s;;

let equiv = function
        '>' -> "++ptr;"
        | '<' -> "--ptr;"
        | '+' -> "++*ptr;"
        | '-' -> "--*ptr;"
        | '.' -> "putchar(*ptr);"
        | ',' -> "*ptr = getchar();"
        | '[' -> "while (*ptr) {"
        | ']' -> "}"
        | _ -> "";;

let translate ch =
        String.iter (fun x ->
                match equiv x with "" -> ()
                | s -> Printf.fprintf ch "      %s\n" s)
        ;;

let transpile input_file output_file =
        let oc = open_out output_file in
        Printf.fprintf oc "%s\n" out_start;
        read_file input_file |> translate oc;
        Printf.fprintf oc "%s" "}";
        close_out oc;;

let compile input_file =
        let output_file = Filename.remove_extension input_file
        in let temp_file = Filename.temp_file output_file ".c"
        in let command = Filename.quote_command "tcc" [temp_file; "-o"; output_file] in
        transpile input_file temp_file;
        (match Sys.command command with 0 -> () | n -> Printf.printf "Error code %d" n);
        Sys.remove temp_file;;

let interpret input_file =
        let content = read_file input_file
        and arr = Array.make 30_000 0
        and c = ref 0 and i = ref 0
        and correct x = x := !x mod 30_000
        in let len = String.length content - 1
        in let find_next s =
                let rec aux j co =
                        if j = len then failwith "Could not find matching ]"
                        else if content.[j] = '[' then aux (j+1) (co+1)
                        else if content.[j] = ']'
                                then if co = 0 then i := j else aux (j+1) (co-1)
                        else aux (j+1) co
                in aux s 0
        and find_prev s =
                let rec aux j co =
                        if j = 0 then failwith "Could not find matching ["
                        else if content.[j] = ']' then aux (j-1) (co+1)
                        else if content.[j] = '['
                                then if co = 0 then i := j else aux (j-1) (co-1)
                        else aux (j-1) co
                in aux s 0
        in let rec aux () =
                if !i>=len then () else begin
                incr i;
                match content.[!i] with
                  '>' -> (incr c; correct c; aux())
                | '<' -> (decr c; correct c; aux())
                | '+' -> (arr.(!c) <- arr.(!c)+1; aux())
                | '-' -> (arr.(!c) <- arr.(!c)-1; aux())
                | '.' -> (arr.(!c) mod 255 |> char_of_int |> print_char; aux())
                | ',' -> (arr.(!c) <- (input_char stdin |> int_of_char); aux())
                | '[' ->
                        if arr.(!c) = 0 then
                                find_next (!i-1)
                        else ();
                        aux ()
                | ']' ->
                        if arr.(!c) <> 0 then
                                find_prev (!i-1)
                        else ();
                        aux ()
                | _ -> aux ()
                end
        in aux ();;

let () =
        let interpreter = ref false
        and transpiler = ref false in
        for i = 1 to Array.length Sys.argv - 1 do
                match Sys.argv.(i) with
                "-i" | "--interpreter" -> interpreter := true
                | "-t" | "--transpiler" -> transpiler := true
                | file -> begin
                                if !interpreter then (interpreter := false;
                                        interpret file;
                                        Printf.printf "\n%s - intepreted\n" file)
                                else if !transpiler then (transpiler := false;
                                        Filename.remove_extension file ^ ".c" |>
                                        transpile file;
                                        Printf.printf "%s - transpiled\n" file)
                                else (compile file; Printf.printf "%s - compiled\n" file)
                        end
        done;;

File: c/bf.c
#include<libgen.h>
#include<stdbool.h>
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<assert.h>
#include<unistd.h>

#define OUT_START "extern int putchar(int);\nextern char getchar();\n\nchar array[30000] = {0}; char *ptr = array;\nint main (int argc, char *argv[]) {"

static void usage(void) {
  fputs("usage: bf [-hit] file1 [file2..]\n\
  -i: interpret\n\
  -t: transpile (to C)\n", stderr);
  exit(EXIT_SUCCESS);
}

inline int mod(int x, int y) {
  return ((x%y) + y)%y;
}

// Thanks https://stackoverflow.com/questions/2736753/how-to-remove-extension-from-file-name
static char *remove_ext (char* myStr, char extSep, char pathSep) {
  char *retStr, *lastExt, *lastPath;

  if (myStr == NULL) return NULL;
  if ((retStr = malloc (strlen (myStr) + 1)) == NULL) return NULL;

  strcpy (retStr, myStr);
  lastExt = strrchr (retStr, extSep);
  lastPath = (pathSep == 0) ? NULL : strrchr (retStr, pathSep);

  if (lastExt != NULL) {
    if (lastPath != NULL) {
      if (lastPath < lastExt) {
        *lastExt = '\0';
      }
    } else {
      *lastExt = '\0';
    }
  }

  return retStr;
}

static char* read_file(const char* const filename) {
	if (filename == NULL)
    return NULL;
  FILE *f = fopen(filename, "r");
	if (f == NULL)
    return NULL;

	assert(!fseek(f, 0, SEEK_END));
	long file_size = ftell(f);
	rewind(f);
	size_t code_size = sizeof(char) * file_size;
	char *code = malloc(code_size);
	if (code == NULL)
		return NULL;
	fread(code, 1, file_size, f);
	assert(!fclose(f));
	return code;
}

static const char* equiv(char c) {
	switch (c) {
    case '>': return "++ptr;";
		case '<': return "--ptr;";
		case '+': return "++*ptr;";
		case '-': return "--*ptr;";
		case '.': return "putchar(*ptr);";
		case ',': return "*ptr = getchar();";
		case '[': return "while (*ptr) {";
		case ']': return "}";
		default:  return "";
	}
}

static void translate(FILE* source, FILE* output) {
	char* a = malloc(17 * sizeof(char));
	char c;
	while((c = fgetc(source)) != EOF) {
		strcpy(a, equiv(c));
		if (strcmp(a, "")) {
			fputs("\t", output);
			fputs(a, output);
			fputs("\n", output);
		}
	}
	free(a);
}

static void interpret(const char* const input) {
	char array[30000] = {0}; char *ptr = array;
	char c;
	for (int i = 0; (c = input[i]) != '\0'; i++) {
		switch (c) {
			case '>': ++ptr; break;
			case '<': --ptr; break;
			case '+': ++(*ptr); break;
			case '-': --(*ptr); break;
			case '.': putchar(*ptr); break;
			case ',': *ptr = getchar(); break;
			case '[':
				if (!*ptr) {
					int loop = 1;
					while (loop > 0) {
						c = input[++i];
						if (c == ']')
							--loop;
						else if (c == '[')
							++loop;
					}
				}
				break;
			case ']':
				if (*ptr) {
					int loop = 1;
					while (loop > 0) {
						c = input[--i];
						if (c == '[')
							--loop;
						else if (c == ']')
							++loop;
					}
				}
				break;
		}
	}
}

static void transpile(char* input, FILE* output) {
	FILE *source = fopen(input, "r");
	if (source == NULL) {
		printf("File %s is not available.\n", input);
		exit(EXIT_FAILURE);
	}
	fputs(OUT_START, output);
	translate(source, output);
	fputs("}", output);
	fclose(source);
	fclose(output);
}

static void compile(char* input) {
	char* output_file = remove_ext(basename(input), '.', '/');
	char* template = malloc(1025 * sizeof(char));

	strcpy(template, output_file);
	strcat(template, "XXXXXX.c");

	int output_fd = mkstemps(template, 2);

	if (output_fd == -1) {
		printf("Error while creating temporary file");
		exit(EXIT_FAILURE);
	}

	FILE* output = fdopen(output_fd, "a");
	transpile(input, output);
	close(output_fd);

	char *command = malloc(sizeof(char) * 2048 + 6);
	sprintf(command, "tcc %s -o %s", template, output_file);

	system(command);
	unlink(template);
	free(output_file);
	free(command);
	free(template);
}

int main(int argc, char *argv[]) {
	bool interpreter = false, transpiler = false;

	for (int i = 1; i<argc; i++) {
		if (!strcmp(argv[i], "-i"))
			interpreter = true;
		else if (!strcmp(argv[i], "-t"))
			transpiler = true;
		else if (!strcmp(argv[i], "-h"))
			usage();
		else {
			char* file = argv[i];
			if (interpreter) {
				interpreter = false;
				char *file_content = read_file(file);
				if (file_content == NULL)
					printf("Couldn't open file");
				else interpret(file_content);
				free(file_content);
        printf("\n%s - Interpreted\n", file);
			} else if (transpiler) {
				transpiler = false;
				char* output_file = strcat(remove_ext(file, '.', '/'), ".c");
				if (output_file == NULL) {
					printf("Error while removing ext\n");
					return 1;
				}
				FILE *target = fopen(output_file, "a");
				transpile(file, target);
				free(output_file);
        printf("%s - Transpiled\n", file);
			} else {
				compile(argv[i]);
        printf("%s - Compiled\n", file);
			}
		}
	}
	return 0;
}
